### ISSUE TITLE
-nan error when training yolo_tiny_pan3

### ISSUE THREAD

**Issue by Hyvenos** [AlexeyAB/darknet#4777]
Hey,

I'm trying to train the yolo_tiny_pan3 you described here: https://github.com/AlexeyAB/darknet/issues/3114#issuecomment-494148968

Here is my cfg file:

`[net]
# Testing
#batch=1
#subdivisions=1
# Training
batch=64
subdivisions=16
width=544
height=544
channels=3
momentum=0.9
decay=0.0005
angle=0
saturation = 1.5
exposure = 1.5
hue=.1
mixup=1
#blur=5

learning_rate=0.02
burn_in=1000
max_batches = 90000

policy=sgdr
sgdr_cycle=1000
sgdr_mult=2
steps=4000,6000,8000,9000
#scales=1, 1, 0.1, 0.1

[convolutional]
batch_normalize=1
filters=16
size=3
stride=1
pad=1
activation=swish

#[convolutional]
#batch_normalize=1
#filters=16
#size=3
#stride=2
#pad=1
#activation=swish

# remove
[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=32
size=3
stride=1
pad=1
activation=swish

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=64
size=3
stride=1
pad=1
activation=swish

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=128
size=3
stride=1
pad=1
activation=swish

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=swish

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=swish

[maxpool]
size=2
stride=1

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=swish

###########

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=swish

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=swish
assisted_excitation=4000

########### to [yolo-3]

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=swish

[upsample]
stride=2

[route]
layers = -1, 8

###########

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=swish

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=swish

########### to [yolo-2]

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=swish

[upsample]
stride=2

[route]
layers = -1, 6

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activation=swish

[convolutional]
batch_normalize=1
filters=128
size=3
stride=1
pad=1
activation=swish

########### features of different layers

[route]
layers=2

[maxpool]
size=16
stride=16

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activation=swish

[route]
layers=4

[maxpool]
size=8
stride=8

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activation=swish

[route]
layers=4

[maxpool]
size=8
stride=4
stride_x=4
stride_y=8

[convolutional]
batch_normalize=1
filters=64
size=1
stride=2
stride_x=2
stride_y=1
pad=1
activation=swish

[route]
layers=4

[maxpool]
size=8
stride=8
stride_x=8
stride_y=4

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
stride_x=1
stride_y=2
pad=1
activation=swish

[route]
layers=4

[maxpool]
size=8
stride=8

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activation=swish

[route]
layers=6

[maxpool]
size=4
stride=4

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activation=swish

[route]
layers=6

[maxpool]
size=4
stride=2
stride_x=2
stride_y=4

[convolutional]
batch_normalize=1
filters=64
size=1
stride=2
stride_x=2
stride_y=1
pad=1
activation=swish

[route]
layers=6

[maxpool]
size=4
stride=4
stride_x=4
stride_y=2

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
stride_x=1
stride_y=2
pad=1
activation=swish

[route]
layers=8

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activation=swish

[route]
layers=10

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activation=swish

[route]
layers=-1, -3, -6, -9, -12, -15, -18, -21, -24, -27

[maxpool]
maxpool_depth=1
out_channels=64
stride=1
size=1

########### [yolo-1]

[upsample]
stride=4

[route]
layers = -1,24

[convolutional]
batch_normalize=1
filters=128
size=3
stride=1
pad=1
activation=swish

[convolutional]
size=1
stride=1
pad=1
filters=35
activation=linear

[yolo]
mask = 0,1,2,3,4
anchors = 8,8, 10,13, 16,30, 33,23,  32,32, 30,61, 62,45, 64,64,  59,119, 116,90, 156,198, 373,326
classes=2
num=12
jitter=.3
ignore_thresh = .7
truth_thresh = 1
iou_normalizer=0.25
cls_normalizer=1.0
iou_loss=giou
scale_x_y = 1.05
random=0

########### [yolo-2]

[route]
layers = -6

[upsample]
stride=2

[route]
layers = -1,19

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=swish

[convolutional]
size=1
stride=1
pad=1
filters=35
activation=linear

[yolo]
mask = 4,5,6,7,8
anchors = 8,8, 10,13, 16,30, 33,23,  32,32, 30,61, 62,45, 64,64,  59,119, 116,90, 156,198, 373,326
classes=2
num=12
jitter=.3
ignore_thresh = .7
truth_thresh = 1
iou_normalizer=0.25
cls_normalizer=1.0
iou_loss=giou
scale_x_y = 1.1
random=0

########### [yolo-3]

[route]
layers = -12

[route]
layers = -1,14

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=swish

[convolutional]
size=1
stride=1
pad=1
filters=28
activation=linear

[yolo]
mask = 8,9,10,11
anchors = 8,8, 10,13, 16,30, 33,23,  32,32, 30,61, 62,45, 59,119,   80,80, 116,90, 156,198, 373,326
classes=2
num=12
jitter=.3
ignore_thresh = .7
truth_thresh = 1
iou_normalizer=0.25
cls_normalizer=1.0
iou_loss=giou
scale_x_y = 1.2
random=0
`

I used the default value provided by this cfg, just removing the antialiasing=1 as you suggested yesterday and changed my class number and filters accordingly.
 I'm using the latest version of this repo (I pulled again like 2 hours ago)
I'm using the .conv I got from `darknet partial path/to/the/cfg  /path/to/the/weights yolo_tiny_pan3.conv.15 15`. Where /path/to/the/weights refers to the weights you provided along with the cfg file.

My issue is that after 10 iterations, I have two iterations with -nan loss, resulting in -nan avg loss until I reload my training using -gpus.
That seems to cause issue with opencv trying to plot a loss graph he can't plot and finally crash after like 6000 iteration.
What is strand is that I got this error after reloading my trying using multi-gpus, at this time I don't avec nan anymore and my loss seems normal.
Here are the error I got:
`OpenCV exception: draw_train_chart() 
OpenCV exception: draw_train_loss()
OpenCV exception: draw_train_loss()
...
OpenCV exception: draw_train_loss()
`
Moreover my map is at 0% for evey weights I could have. Thus for this point I don't know if it's normal or if I just have to wait for more iterations.

If someone have an idea or have already encountered such issue, I would be grateful!

**Comment by AlexeyAB** [comment]
Use `learning_rate=0.001`
And set `mosaic=1` instead of ` mixup=1`

**Comment by Hyvenos** [comment]
I tried your modifications. It takes more time to get a -nan value but I still have it. What is surprising is that the loss is actually decreasing, and then I got like 3 -nan loss, I don't know why, and it starts outputing correct decreasing values like any others of my trainings.
The training is still running so I'll update my answer when it will be more advanced, but for the 2000.weights the map is still at 0 and I still get `OpenCV exception: draw_train_loss()` on the error output, thus it doesn't seems to lead to a crash of the darknet process, as I could have my previous training (with learning_rate=0.02 and mixup=1) running from beginning to the end.

**Comment by AlexeyAB** [comment]
@Hyvenos 

* SHow chart.png
* check files bad.list bad_label.list

**Comment by Hyvenos** [comment]
Here is my chart.png, as you can see only some first iterations are plotted:
![chart](https://user-images.githubusercontent.com/31487281/73540532-74633880-4430-11ea-8f7b-37de1393051b.png)

I don't have a bad_label.list and my bad.list only contains data/labels/*.png

**Comment by AlexeyAB** [comment]
> I don't have a bad_label.list and my bad.list only contains data/labels/*.png

Check that you have these png files in your train.txt. If yes, then do you have these files on your disk?

**Comment by Hyvenos** [comment]
I confirm that, after 40 000 iterations my map is still at zero and my chart.png is still the same.
No I don't have these files in my train.txt and I also don't have them on my disk. But for whatever reason, for every training I do, my bad.list is filled with these files.
I though it was some kind of relicate from the COCO challenge or some stuff like that, that appears when running either detector train/map/test.
In fact I'm doing these in parallel as my training is running, so I can't ensure my bad.list come from detector train command, but I can check if it's useful for you.
By the way in which folder is it supposed to save bad.list? It seems that it's saved in the folder in which the terminal where I launched the command is. But I'm not sure.

**Comment by AlexeyAB** [comment]
> By the way in which folder is it supposed to save bad.list? It seems that it's saved in the folder in which the terminal where I launched the command is. But I'm not sure.

Yes.

Delete these files and train again with latest Darknet.

If avg_loss is Nan, then you must stop training.

Check your dataset by training with flag `-show_imgs` do you see correct bboxes?

**Comment by Hyvenos** [comment]
I tried your suggestion on my personal computer, as I can't use show_img on my server.
For what I see, it works well: the saved imgs show me correct boxes and I didn't see any -nan, thus I only tried for 40 iterations. Have you changed something?
 I will tried further this morning.

**Comment by Hyvenos** [comment]
I let the training run until iterations 5000 on my training server, and it works like a charms. No more -nan, the loss decrease effectively (it was stuck around 2 on my previous training) and the map I processed where above 0. So I can say without doubt that it works for me now, but I don't understand what changed.

Looking on my maps' logfiles, I saw an error occuring sometimes:
cuDNN status Error in: file: ./src/convolutional_kernels.cu : () : line: 543 : build time: Jan 27 2020 - 08:55:26 cuDNN Error: CUDNN_STATUS_BAD_PARAM
But this doesn't happen to all the map processed so I don't think it is related to the failure on my map processing.

Do you think I can replace again the mosaic=1 by mixture=1 and add the antialiasing=1 you made me removed on the maxpool layers? Would it help to improve the detections? I suppose antialiasing=1 improve the quality of the maxpooling

**Comment by AlexeyAB** [comment]
mosaic is better than mixup.
antialiasing doesn't give any advantage.

**Comment by Hyvenos** [comment]
Ok perfect, I will close this issue since everything is working.
Thanks for your help!
